### PR TITLE
ipn/ipnlocal: name the unlockOnce type, plumb more, add Unlock method

### PR DESCRIPTION
**Pull request by bradfitz**
This names the func() that Once-unlocked LocalBackend.mu. It does so
both for docs and because it can then have a method: Unlock, for the
few points that need to explicitly unlock early (the cause of all this
mess). This makes those ugly points easy to find, and also can then
make them stricter, panicking if the mutex is already unlocked. So a
normal call to the func just once-releases the mutex, returning false
if it's already done, but the Unlock method is the strict one.

Then this uses it more, so most the b.mu.Unlock calls remaining are
simple cases and usually defers.

Updates #11649

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -4451,9 +4457,30 @@ func (b *LocalBackend) stateMachine() {
...
+	}
+}
+
+// unlockOnce is a func that unlocks only b.mu the first time it's called.
+// Therefore it can be safely deferred to catch error paths, without worrying
+// about double unlocks if a different point in the code later needs to explicitly
+// unlock it first as well. It reports whether it was unlocked.
+type unlockOnce func() bool
+
+// Unlock unlocks the LocalBackend.mu.
```
Comment (maisem) [inline ipn/ipnlocal/local.go:4478]: This needs more docs. Or at least a different name, like UnlockEarly

Comment (maisem) [reply ipn/ipnlocal/local.go:4478]: Like what property does it provide by panicing

Comment (bradfitz) [reply ipn/ipnlocal/local.go:4478]: Documented, renamed.

File: ipn/ipnlocal/local.go [hunk 2]
```diff
@@ -4451,9 +4457,48 @@ func (b *LocalBackend) stateMachine() {
...
+	}
+}
+
+// unlockOnce is a func that unlocks only b.mu the first time it's called.
+// Therefore it can be safely deferred to catch error paths, without worrying
+// about double unlocks if a different point in the code later needs to explicitly
+// unlock it first as well. It reports whether it was unlocked.
+type unlockOnce func() bool
+
+// UnlockEarly unlocks the LocalBackend.mu. If panics if u returns false,
```
Comment (maisem) [inline ipn/ipnlocal/local.go:4478]: It panics

**Review by maisem — APPROVED** [review]
Thanks! Much easier to understand!

**Pull request merged**